### ISSUE TITLE
appcrash on "Remove All Hooks" with CSharpConsole 

### ISSUE THREAD

**Issue by nblog** [nektra/Deviare2#50]
appcrash in after hooked, remove the hook again.
1. select any process and select kernel32.dll!CreateFile (maybe any functions)
![Snipaste_2019-12-13_20-05-10](https://user-images.githubusercontent.com/10218627/70799178-e41b5680-1de3-11ea-8229-665c8ed52c29.png)
2. click "Remove All Hooks"
![Snipaste_2019-12-13_20-05-10](https://user-images.githubusercontent.com/10218627/70799275-1cbb3000-1de4-11ea-817c-95a72c9bba88.png)
3. let the program call
![Snipaste_2019-12-13_20-05-10](https://user-images.githubusercontent.com/10218627/70799405-81768a80-1de4-11ea-9e35-e517a7b1c34a.png)

The program has crashed.

After I tried to track it, I found that
[HookEngine.cpp#L806](https://github.com/nektra/Deviare2/blob/master/Source/Engine/Agent/HookEngine.cpp#L806)
 CNktAutoFastMutex cLock(this)  // crash point

**Comment by mxmauro** [comment]
Hi @nblog the line you mention does not match the latest code on GitHub. Are you using the latest version?

If yes, can you upload a crash dump + pdb files to analyze the scenario?

Regards,
Mauro.

**Comment by nblog** [comment]
It seems that there is no crash dump, Sorry, I don't know how to use it.

But I tested the latest version, and releases version.
<del>
I analyzed it again and found that
CSharpConsole clicked  "remove all hooks"
Source\Engine\Agent\AgentMgr.cpp -> "CDvAgentMgr::OnEngMsg_EnableHook", 
Not work.
</del>

**Comment by nblog** [comment]
Maybe I described it wrong, Crash is the target program.

**Comment by nblog** [comment]
If my analysis is correct
click "Remove All Hooks"
CDvAgentMgr::TAC_OnEngineMessage -> dwMsgCode == NKT_DV_TMSG_CODE_RemoveHook
CNktDvHookEngine::Unhook ->
cHookLib.EnableHook(aHookLibItemsList, nTempHookEntriesCount, FALSE); ->
CNktHookLib::EnableHookCommon ->

After entering, it was not implemented
NktHookLibHelpers::WriteMem(lpHookEntry->cProcEntry, lpHookEntry->lpInjData + 1, &nVal, 1);
or
_InterlockedAnd((LONG volatile *)(lpHookEntry->lpInjData), 0xFFFF00FF);

If I make other progress
I will tell you in time.

thanks again.

**Comment by nblog** [comment]
I think I have discovered the cause of the problem.

Deviare2/Source/Engine/Agent/HookEngine.cpp
function:  CNktDvHookEngine::Unhook

Before calling "cHookLib.EnableHook (aHookLibItemsList, nTempHookEntriesCount, FALSE);"
calling "cHooksList.Remove(lpHookEntry);"
At this point the hook object no longer exists.

"cHookLib.EnableHook", Will no longer take effect.

After I tried to modify, it started working properly.
![Snipaste_2019-12-13_22-07-44](https://user-images.githubusercontent.com/10218627/70805894-01a4ec00-1df5-11ea-82b5-314405967da5.png)

Although the hook was not actually removed, it was banned.
After the program is closed, it will be truly removed.
